src/time.rs

```diff
@@ -0,0 +1,189 @@
+    pub fn from_millis(millis: i64) -> Duration {
+        Duration { millis }
+    }
+
+    /// The fractional number of milliseconds in this `Duration`.
+    pub fn millis(&self) -> i64 {
+        self.millis % MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds in this `Duration`.
+    pub fn secs(&self) -> i64 {
+        self.millis / MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds in this `Duration`.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl ops::Add<Duration> for Duration {
+    type Output = Duration;
+
+    fn add(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis + rhs.total_millis())
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Duration;
+
+    fn sub(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis - rhs.total_millis())
+    }
+}
+
+impl ops::Mul<Duration> for Duration {
+    type Output = Duration;
+
+    fn mul(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis * rhs.total_millis())
+    }
+}
+
+impl ops::Div<Duration> for Duration {
+    type Output = Duration;
+
+    fn div(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis / rhs.total_millis())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use core::i64;
+    use super::*;
+
+    #[test]
+    fn test_instant_ops() {
+        // std::ops::Add
+        assert_eq!(Instant::from_millis(4) + Duration::from_millis(6), Instant::from_millis(10));
+        // std::ops::Sub
+        assert_eq!(Instant::from_millis(7) - Duration::from_millis(5), Instant::from_millis(2));
+    }
+
+    #[test]
+    #[should_panic(expected = "attempt to add with overflow")]
+    fn test_add_with_overflow() {
+        Instant::from_millis(i64::MAX) + Duration::from_millis(6);
+    }
+
+    #[test]
+    fn test_instant_getters() {
+        let instant = Instant::from_millis(5674);
+        assert_eq!(instant.secs(), 5);
+        assert_eq!(instant.millis(), 674);
+        assert_eq!(instant.total_millis(), 5674);
+    }
+
+    #[test]
+    fn test_instant_display() {
+        assert_eq!(format!("{}", Instant::from_millis(5674)), "5.674s");
+        assert_eq!(format!("{}", Instant::from_millis(5000)), "5.0s");
+    }
+
+    #[test]
+    fn test_duration_ops() {
+        // std::ops::Add
+        assert_eq!(Duration::from_millis(40) + Duration::from_millis(2), Duration::from_millis(42));
+        // std::ops::Sub
+        assert_eq!(Duration::from_millis(555) - Duration::from_millis(42), Duration::from_millis(513));
+        // std::ops::Mul
+        assert_eq!(Duration::from_millis(13) * Duration::from_millis(22), Duration::from_millis(286));
+        // std::ops::Div
+        assert_eq!(Duration::from_millis(53) / Duration::from_millis(4), Duration::from_millis(13));
+    }
+
+    #[test]
+    #[should_panic(expected = "attempt to multiply with overflow")]
```

Comment:
This test and the other overflow test above would fail on `cargo test --release` since that doesn't include overflow checks and you don't explicitly use checked operations. I suggest simply removing them, and letting the default Rust arithmetic rules apply to time manipulation, since 64-bit signed overflows and underflows are really not important in this context.

There is of course the important exception of `Duration - Duration`, which is an unsigned underflow, and should be tested.

Reply:
Right. I'll update this.